beam-lib/src/http_util.rs

```diff
@@ -1,21 +1,244 @@
+        }
+    }
+}
+
+impl BeamClient {
+    /// Create a beam client based on the Authorization credentials and the beam proxy url
+    pub fn new(app: &AddressingId, key: &str, beam_proxy_url: Url) -> Self {
+        let default_headers = [
+            (header::AUTHORIZATION, HeaderValue::from_bytes(format!("ApiKey {app} {key}").as_bytes()).expect("This is a valid header value"))
+        ].into_iter().collect();
+        Self {
+            client: Client::builder().default_headers(default_headers).build().expect("Client should always build"),
+            beam_proxy_url
+        }
+    }
+
+    /// Construct a beam client from a [`reqwest::Client`]
+    /// # Note:
+    /// This expects the user to have set the default headers for authorization accordingly.
+    /// If you don't need any special client configuration use [`BeamClient::new`].
+    pub fn from_client(client: Client, beam_proxy_url: Url) -> Self {
+        Self { client, beam_proxy_url }
+    }
+
+    /// Poll beam tasks using the `filter=todo` option and the given blocking options.
+    /// The generic Parameter T represents the task body type that the requests are expected to have.
+    pub async fn poll_pending_tasks<T: DeserializeOwned>(&self, blocking: &BlockingOptions) -> Result<Vec<TaskRequest<T>>> {
+        let url = self.beam_proxy_url
+            .join(&format!("/v1/tasks?filter=todo&{}", blocking.to_query()))
+            .expect("The proxy url is valid");
+        let response_result = self.client
+            .get(url)
+            .send()
+            .await;
+        let response = match response_result {
+            Ok(res) => res,
+            Err(e) => return if e.is_timeout() {
+                Ok(Vec::with_capacity(0))
+            } else {
+                Err(e.into())
+            },
+        };
+        match response.status() {
+            StatusCode::GATEWAY_TIMEOUT => Ok(Vec::with_capacity(0)),
+            StatusCode::OK | StatusCode::PARTIAL_CONTENT => Ok(response.json().await?),
+            status => Err(BeamError::UnexpectedStatus(status))
+        }
+    }
+
+    /// Poll beam results for a given task id using the given blocking options.
+    /// The generic Parameter T represents the result body type that the requests are expected to have.
+    pub async fn poll_results<T: DeserializeOwned>(&self, task_id: &MsgId, blocking: &BlockingOptions) -> Result<Vec<TaskResult<T>>> {
+        let url = self.beam_proxy_url
+            .join(&format!("/v1/tasks/{task_id}/results?{}", blocking.to_query()))
+            .expect("The proxy url is valid");
+        let response_result = self.client
+            .get(url)
+            .send()
+            .await;
+        let response = match response_result {
+            Ok(res) => res,
+            Err(e) => return if e.is_timeout() {
+                Ok(Vec::with_capacity(0))
+            } else {
+                Err(e.into())
+            },
+        };
+        match response.status() {
```

Comment:
Currently, branching based on `response.is_success()` might be clearer.

Reply:
I disagree.
1. If we chose to add different errors in the future it will be a match anyway
2. I chose to handle a gateway timeout as returning no elements because the application logic will almost always treat these cases the same way so I need to check for the `GATEWAY_TIMEOUT` status.
3. I think it's a good idea to be really strict with what statuses we treat as valid as it will also help us catch breaking changes in beam.

Reply:
Fair points. My suggestion was based on seeing all successful status codes we process or-ed together. Your points, especially 2 and 3, are good reasons, tho.